### ISSUE TITLE
ansible-lint erroneously detects pipes in shell when that is template

### ISSUE THREAD

**Issue by ignatenkobrain** [ansible/ansible-lint#544]
# Issue Type
- Bug report

# Ansible and Ansible Lint details

```
⋊> ~/P/g/ansible on master ◦ ansible --version                                                                                                                              13:30:37
ansible 2.8.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/brain/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.4rc2 (default, Jul  3 2019, 01:30:40) [GCC 9.1.1 20190605 (Red Hat 9.1.1-2)]
⋊> ~/P/g/ansible on master ◦ ansible-lint --version                                                                                                                         13:30:54
ansible-lint 4.1.0
```

- ansible installation method: OS package
- ansible-lint installation method: OS package

# Desired Behaviour

ansible-lint should not find this false positive…

# Actual Behaviour (Bug report only)

```yaml
---
- name: Test
  hosts: localhost
  gather_facts: false
  tasks:
  - name: Test
    shell: echo
      {% for i in range(nameservers | length) %}
      {{ nameservers[i] }}
      {% endfor %}
    changed_when: false
  vars:
    nameservers:
    - 8.8.8.8
    - 4.4.4.4
```

And ansible-lint returns:

```
[306] Shells that use pipes should set the pipefail option
test.yml:6
Task/Handler: Test
```

I did not check code, but I think it just checks whether pipe is present in shell string, but shell is actually jinja2 template, so in this case it is false-positive.

**Comment by willthames** [comment]
That code should fail a "no `{%`" in playbooks check!

You're quite correct though, it's a simple regex. Checking templates is hard, not sure what the easy fix to this one is.

Checking whether the pipe is inside templating might be the simplest - it wouldn't fix your scenario but it would fix `echo {{ nameservers | length }}`.

**Comment by ignatenkobrain** [comment]
Hello @willthames,

> That code should fail a "no {%" in playbooks check!

Can you give me link to this definition of a check please? I'm curious what is the better way to construct commandline out of the template.

**Comment by willthames** [comment]
@ignatenkobrain there is no such check, I just think there ought to be!

It really depends what you're trying to do - in your case it just looks like you're doing `echo {{ nameservers | join(" ") }}` but I assumed that was a simple example that demonstrated this issue. 

Put it this way, in the 7 years of using ansible, I've always managed to avoid `{%` in playbooks and roles (they're fine in templates when used appropriately). In the absolute worst case I'd convert the shell script to a template file

**Comment by HugoPoi** [comment]
I confirm the issue with the 306 rule. It take `|` in `{{}}` for pipe in the shell command itself, there is something wrong with this check.

**Comment by EpiqSty** [comment]
Same issue with fake alert, but a bit different case:
```
    - name: get difference between backups
      shell: diff -I '^## Last [change|commit]' {{ prev_config }} {{ working_configuration }}
      when: prev_config is defined
      register: diff_result
      failed_when: diff_result.rc > 1
```

As you could see, pipe here should be ignored, as it's just a part of argument, string, not a shell pipe.

**Comment by aureliojargas** [comment]
The issue with the pipe `|` inside `{{...}}` seems to be solved now. I've tested it locally in ansible-lint 4.2.0 and it worked as expected.

The current code is using a regex to "unjinja" before checking:

https://github.com/ansible/ansible-lint/blob/c59afffbc6975e6fd97330b69a4b8c640159978f/lib/ansiblelint/rules/ShellWithoutPipefail.py#L34-L35

https://github.com/ansible/ansible-lint/blob/c59afffbc6975e6fd97330b69a4b8c640159978f/lib/ansiblelint/__init__.py#L46-L47

But if you have a pipe inside `{% ... %}`, it's still a false positive for this check.
